File: src/main/java/core/procedures/generic/Count.java
package core.procedures.generic;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.Collection;
import java.util.Map;

public class Count extends AFn {

  public Count() {
    super(new FnArgsBuilder().min(1).max(1).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "count";
  }

  @Override
  public Integer apply1(Object arg) {
    if (arg instanceof Map) {
      return ((Map) arg).size();
    } else if (arg instanceof Map.Entry) {
      return 2;
    } else if (arg instanceof Collection) {
      return ((Collection)arg).size();
    } else if (arg instanceof CharSequence) {
      return ((CharSequence) arg).length();
    }
    throw new WrongTypeException(getName(), "List or Map or Vector or Set or String", arg);
  }
}

File: src/main/java/core/procedures/generic/Count.kt
package core.procedures.generic

import core.exceptions.WrongTypeException
import core.procedures.AFn

open class Count : AFn<Any?, Int>(name = "count", isPure = true, minArgs = 1, maxArgs = 1) {

    override operator fun invoke(arg: Any?) = when (arg) {
        is Map.Entry<*, *> -> 2
        is Map<*, *>       -> arg.size
        is Collection<*>   -> arg.size
        is CharSequence    -> arg.length
        else               -> throw WrongTypeException(name, "List or Map or Vector or Set or String", arg)
    }
}
